src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -58,13 +58,90 @@ class PredictionCache {
   std::shared_ptr<metrics::RatioCounter> hit_ratio_;
 };
 
-template <typename Scheduler>
+struct DeadlineCompare {
+  bool operator()(const std::pair<Deadline, PredictTask>& lhs,
+                  const std::pair<Deadline, PredictTask>& rhs) {
+    return lhs.first > rhs.first;
+  }
+};
+
+// thread safe model queue
+class ModelQueue {
+ public:
+  ModelQueue() : queue_(ModelPQueue{}) {}
+
+  // Disallow copy and assign
+  ModelQueue(const ModelQueue&) = delete;
+  ModelQueue& operator=(const ModelQueue&) = delete;
+
+  ModelQueue(ModelQueue&&) = default;
+  ModelQueue& operator=(ModelQueue&&) = default;
+
+  ~ModelQueue() = default;
+
+  void add_task(PredictTask task) {
+    std::unique_lock<std::mutex> l(queue_mutex_);
+    Deadline deadline = std::chrono::high_resolution_clock::now() +
+        std::chrono::microseconds(task.latency_slo_micros_);
```

Comment:
Yeah as I mentioned above, let's set the deadline for the query once when it first arrives, then use it throughout execution.